ensure narinfo.Check works on uncompressed NARs

manveru opened the pull request:
Without this, using `Check` would result in an error:
```
panic: cannot handle unexported field at {*hash.Hash}.hash:
        "github.com/nix-community/go-nix/pkg/hash".Hash
consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported
```
While this solution may not be the most efficient, I'm open to suggestions and just wanted to quickly fix this for myself.

flokli commented:
Thanks!

We should probably use a custom [Comparer](https://github.com/google/go-cmp/blob/v0.5.9/cmp/options.go#L344), that compares the multihash representation (`[]byte`). It should be faster than converting to a string representation and comparing that.

Also, as the FileHash and FileSize fields are optional, we might want to flip the logic in the Check function a bit. I think it's too restrictive and might fail on some NARInfo files that don't have the fields set.

Would you be open to updating the PR to reflect that?

manveru commented:
I don't think using `go-cmp` buys us anything other than adding a dependency. So I used a simple `bytes.Equal` instead.
Also to make the error message nicer, I added the `Hash.HashTypeString` function and simplified the code setting the `hashtypeToNixHashString` a little bit.

manveru commented:
Not sure why the fixture test fails... I tried running `./build-fixtures.go` and it didn't produce any diff for me.

flokli commented:
Only got to dig into this now - it seems the output produced by Nix changed (now keys are sorted alphabetically, not arbitrarly).

PR is up at https://github.com/nix-community/go-nix/pull/100, once that's in we should be able to rebase this and not get CI failures.

flokli commented:
Erf, I tried to rebase, but did forget to include your commit before pushing, and GH closed the PR, not allowing me to reopen :-/

I cherry-picked it to https://github.com/nix-community/go-nix/pull/101
